### PR TITLE
feat: Allow fine-tuning of K8s rest client connection properties

### PR DESCRIPTION
**Pull request by jannfis**
This adds some under-the-hood options for fine tuning certain connection parameters of the Kubernetes REST client used by the application controller and API server, using environment variables.

The main motivation behind this is the ability to stabilize operations to clusters at the far edge, which may be on very high latency and very low bandwidth links, and the default timeouts may not be sufficient for proper operations.

It introduces the following environment variables to change the properties from their defaults:

* `ARGOCD_K8S_TCP_TIMEOUT` - (duration) allows to set the general timeout for TCP connections to the K8s API, and also sets the server-side timeout to the same value
* `ARGOCD_K8S_TCP_KEEPALIVE` - (duration) allows to set the interval at which TCP keep alive packets will be sent to the K8s API
* `ARGOCD_K8S_TLS_HANDSHAKE_TIMEOUT` - (duration) allows to set the timeout for waiting on the TLS handshake to complete with the K8s API 
* `ARGOCD_K8S_TCP_IDLE_TIMEOUT` - (duration) allows to set the timeout for closing idle TCP connections to the K8s API server

Signed-off-by: jannfis <jann@mistrust.net>

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [ ] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

**Review by kevchu3 — COMMENTED** [review]
@jannfis and I performed some testing at a customer site with changing these values, and were successful.  While not disclosing the values we tuned, I thought it'd be good to provide the published defaults here.  This should probably make it into documentation somewhere down the road as well.

Env Var | Value type | Default | Description
-- | -- | -- | --
ARGOCD_K8S_TCP_TIMEOUT | Duration | 30s | General timeout for TCP connections to Kubernetes API
ARGOCD_K8S_TCP_KEEPALIVE | Duration | 30s | Interval at which TCP keepalive probes should be sent
ARGOCD_K8S_TLS_HANDSHAKE_TIMEOUT | Duration | 10s | Timeout for establishing TLS connections to Kubernetes API
ARGOCD_K8S_TCP_IDLE_TIMEOUT | Duration | 5m | Timeout for idle TCP connections

**Review by alexmt — APPROVED** [review]
LGTM

File: pkg/apis/application/v1alpha1/cluster_constants.go [hunk 1]
```diff
@@ -32,26 +46,30 @@ var (
 
 	// K8sMaxIdleConnections controls the number of max idle connections in K8s REST client HTTP transport
 	K8sMaxIdleConnections = 500
+
+	// K8sTLSHandshakeTimeout defines the maximum duration to wait for a TLS handshake to complete
+	K8sTLSHandshakeTimeout = 10 * time.Second
```
Comment (alexmt) [inline pkg/apis/application/v1alpha1/cluster_constants.go:51]: We could inline the init method. It makes it a bit easier to understand what default it.

```
K8sTLSHandshakeTimeout = env.ParseDurationFromEnv(EnvK8sTLSHandshakeTimeout, 10 * time.Second, 0, math.MaxInt32)
```

Totally nice to have. Feel free to ignore suggestion

**Comment by jannfis** [comment]
Cherry-picked this commit into `release-2.3` branch.

**Pull request merged**